Sources/Utility/Process.swift

```diff
@@ -0,0 +1,295 @@
+        defer { for case let arg? in argv { free(arg) } }
+        
+        let env: [UnsafeMutablePointer<CChar>?] = environment.map{ "\($0.0)=\($0.1)".withCString(strdup) }
+        defer { for case let arg? in env { free(arg) } }
+
+        rv = posix_spawnp(&processID, argv[0], &fileActions, &attributes, argv + [nil], env + [nil])
+
+        posix_spawn_file_actions_destroy(&fileActions)
+        posix_spawnattr_destroy(&attributes)
+
+        guard rv == 0 else {
+            throw SystemError.posix_spawn(rv, arguments)
+        }
+    }
+
+    /// Blocks the calling process until the subprocess finishes execution.
+    ///
+    /// - Throws: Process.Error
+    public func waitUntilExit() throws {
+        assert(launched, "The process is not yet launched.")
+        // FIXME: We should probably read the output from somewhere else, maybe in a seperate thread?
+        if let outputStream = outputStream {
+            try readOutput(on: outputStream)
+        }
+        var exitStatus: Int32 = 0
+        var result = waitpid(processID, &exitStatus, 0)
+        while (result == -1 && errno == EINTR) {
+            result = waitpid(processID, &exitStatus, 0)
+        }
+        if result == -1 {
+            throw SystemError.waitpid(errno)
+        }
+
+        if exitStatus & 0x7f == 0 {
+            self.returnCode = (exitStatus >> 8) & 0xff
+        } else {
+            // We received a signal.
+            // FIXME: We may need a better way to represent this.
+            signalled = true
+            self.returnCode = exitStatus
+        }
+    }
+
+    /// Reads the output from save output pipes.
+    ///
+    /// - Throws: Process.Error
+    private func readOutput(on stream: OutputByteStream) throws {
+        assert(saveOutputPipe != [0, 0], "Pipes not initialized.")
+        // Close the write end of the output pipe.
+        let rv = close(saveOutputPipe[1])
+        guard rv == 0 else {
+            throw SystemError.close(rv)
+        }
+
+        // Read all of the data from the output pipe.
+        let N = 4096
+        var buf = [Int8](repeating: 0, count: N + 1)
+        
+        loop: while true {
+            let n = read(saveOutputPipe[0], &buf, N)
+            switch n {
+            case  -1:
+                if errno == EINTR {
+                    continue  // try again!
+                } else {
+                    throw SystemError.read(errno)
+                }
+            case 0:
+                break loop
+            default:
+                buf[n] = 0 // must null terminate
+                if let str = String(validatingUTF8: buf) {
+                    stream <<< str
+                } else {
+                    throw SystemError.popen(EILSEQ, arguments[0])
+                }
+            }
+        }
+
+        // Close the read end of the output pipe.
+        close(saveOutputPipe[0])
+    }
+
+    public func interrupt() {
```

Comment:
One function to just send any signal seems reasonable here, versus separate functions.